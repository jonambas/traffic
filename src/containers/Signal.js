import React, { Component, createContext } from 'react';
import { initialPattern, pairs, pattern, durations } from '../config';
import _ from 'lodash';

const defaultContext = {
  signal: {},
  activePairIndex: 0,
  activePatternIndex: 0
};

export const SignalContext = createContext(defaultContext);

class SignalProvider extends Component {
  state = defaultContext;

  componentDidMount() { 
    this.updateSignal();
  }

  updateSignal = (nextPairIndex = 0, nextPatternIndex = 0) => {
    const activePair = pairs[nextPairIndex];
    const nextPattern = pattern[nextPatternIndex];
    const roads = _.flatten(pairs);

    const state = {
      activePairIndex: nextPairIndex,
      activePatternIndex: nextPatternIndex
    };
    
    /**
     * Describes road state, keyed by road name.
     * @example
     *  east: {left: false, straight: true}
     *  north: {left: false, straight: false}
     *  south: {left: false, straight: false}
     *  west: {left: false, straight: true}
     */
    const signal = roads.reduce((acc, road) => ({
      ...acc,
      [road]: _.includes(activePair, road) ? nextPattern : initialPattern
    }), {});

    this.setState({ ...state, signal }, this.next);
  }

  /**
   * Increments activePairIndex and activePatternIndex
   */
  next = () => {
    setTimeout(() => {
      const { activePairIndex, activePatternIndex } = this.state;
      const nextPatternIndex = activePatternIndex + 1;
      let nextPairIndex = activePairIndex;
      
      if (nextPatternIndex >= pattern.length) {
        nextPairIndex = activePairIndex + 1;
      }
      
      this.updateSignal(nextPairIndex % pairs.length, nextPatternIndex % pattern.length);
    }, durations.patternDuration)
  }

  /**
   * Reshapes data from Queue Context
   * This probably should not be in this component
   */
  getValues = () => {
    const { roadQueue } = this.props;
    const { signal } = this.state;
    const mergedData = {};
    
    _.forEach(signal, (value, key) => {
      mergedData[key] = { ...value, lanes: roadQueue[key] }
    });

    return mergedData;
  }

  render() {
    return (
      <SignalContext.Provider value={this.getValues()}>
        {this.props.children}
      </SignalContext.Provider>
    )
  }
}

export default SignalProvider;