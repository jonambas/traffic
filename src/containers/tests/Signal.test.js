import React from 'react';
import Signal from '../Signal';
import { shallow } from 'enzyme';

describe('Signal Provider', () => {
  let wrapper;
  let next;

  const roadQueueMock = {
    north: [0,1],
    south: [0,2],
    east: [0,3],
    west: [0,4]
  }

  beforeEach(() => {
    global.setTimeout = jest.fn((fn) => next = fn);
    wrapper = shallow(<Signal roadQueue={roadQueueMock}>child</Signal>);
  });

  it('should correctly step through patterns', async () => {
    for (let i = 0; i < 4; i++) {
      await next();  
      expect(wrapper.prop('value')).toMatchSnapshot(); 
    }
  });
});