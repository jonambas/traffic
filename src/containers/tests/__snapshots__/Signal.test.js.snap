// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signal Provider should correctly step through patterns 1`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": false,
    "straight": false,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": false,
    "straight": true,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": false,
    "straight": true,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": false,
    "straight": false,
  },
}
`;

exports[`Signal Provider should correctly step through patterns 2`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": true,
    "straight": false,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": false,
    "straight": false,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": false,
    "straight": false,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": true,
    "straight": false,
  },
}
`;

exports[`Signal Provider should correctly step through patterns 3`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": false,
    "straight": true,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": false,
    "straight": false,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": false,
    "straight": false,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": false,
    "straight": true,
  },
}
`;

exports[`Signal Provider should correctly step through patterns 4`] = `
Object {
  "east": Object {
    "lanes": Array [
      0,
      3,
    ],
    "left": false,
    "straight": false,
  },
  "north": Object {
    "lanes": Array [
      0,
      1,
    ],
    "left": true,
    "straight": false,
  },
  "south": Object {
    "lanes": Array [
      0,
      2,
    ],
    "left": true,
    "straight": false,
  },
  "west": Object {
    "lanes": Array [
      0,
      4,
    ],
    "left": false,
    "straight": false,
  },
}
`;
